src/webview/js/js.js

```diff
@@ -466,31 +487,57 @@ const scrollToPreserve = (msgId: number, prevBoundTop: number) => {
   window.scrollBy(0, newBoundRect.top - prevBoundTop);
 };
 
-const appendAuthToImages = auth => {
-  const imageTags = document.getElementsByTagName('img');
-  arrayFrom(imageTags).forEach(img => {
-    if (!img.src.startsWith(auth.realm)) {
-      return;
-    }
+/**
+ * Rewrite the source URLs of all relative-URL <img> tags beneath the specified
+ * parent element to have an absolute path based on the Zulip realm.
+ *
+ * DEPRECATED: If no parent element is specified, we transform every <img> in
+ * the entire document.
+ */
+const rewriteImageUrls: (auth: Auth, parent?: Element | Document) => void = (() => {
+  // List of routes which accept the API key appended as a GET parameter.
+  const inlineApiRoutes: string[] = ['/user_uploads/', '/thumbnail?', '/avatar/'];
+
+  return (auth: Auth, parent: Element | Document = document) => {
+    const realm: URL = new URL(auth.realm);
+    const imageTags = parent.getElementsByTagName('img');
+
+    arrayFrom(imageTags).forEach(img => {
+      // Get the raw `src` value from the DOM. (We can't easily use `img.src`,
+      // since it's absolutized by the browser.)
+      //
+      /* $FlowFixMe: upstream `getNamedItem` return type should be `Attr | null` */
+      const actualSrc: string | null = img.attributes.getNamedItem('src')?.value;
+
+      // Skip completely sourceless elements: they're someone else's problem.
+      if (actualSrc === null) {
+        return;
+      }
 
-    // This unusual form of authentication is only accepted by the server
-    // for a small number of routes.  Rather than append the API key to all
-    // kinds of URLs on the server, do so only for those routes.
-    const srcPath = img.src.substring(auth.realm.length);
-    if (
-      !(
-        srcPath.startsWith('/user_uploads/')
-        || srcPath.startsWith('/thumbnail?')
-        || srcPath.startsWith('/avatar/')
-      )
-    ) {
-      return;
-    }
+      // Compute the absolute URL as though this page were the basis. (This
+      // should be equivalent to `img.src`, and may even be identical.)
+      const deducedSrc: string = new URL(actualSrc).href;
+
+      // Compute the absolute URL as though `auth.realm` were the basis.
+      let fixedSrc: string = new URL(actualSrc, realm).href;
+
+      // If the corrected URL is on this realm...
+      if (!fixedSrc.startsWith(auth.realm)) {
```

Comment:
Wait, is this the right way around?

Reply:
Also I'm not sure offhand whether `auth.realm` ends in a slash. If not, then this string-prefix check can go wrong.

In the next line, we'll check if `actualSrc` matches any of several patterns, each of which starts with `/`. So we could actually just drop this outer check.

Reply:
It's not, no. I need to write actual tests for this function.

`auth.realm` does not end in a slash unless something has already gone wrong. See `fixRealmUrl` in `url.js`, and its tests in `url-test.js`. (Also, this is exactly what the old code does.)

Oof. We can't drop this outer check, as the server may have sent us an image URL which is already explicitly absolutized to `auth.realm`, but does not have the API key attached. Going on to compare the raw `src` to strings which begin with `'/'` is a bug.

Reply:
> Oof. We can't drop this outer check, as the server may have sent us an image URL which is already explicitly absolutized to `auth.realm`, but does not have the API key attached.

Hmm, yeah, that probably can happen.

That isn't helped by this outer check, though, right? Ah, I guess you mean something like: we need to first check the scheme+authority against `realm` (i.e. a version of this outer check), and then check the *path* -- pulled out from the fixed URL -- against the patterns.